src/core/reaction_ensemble.cpp

```diff
@@ -69,26 +78,22 @@ void EnergyCollectiveVariable::load_CV_boundaries(
     while (iss >> value) {
       values.push_back(value);
     }
-    m_current_wang_landau_system.min_boundaries_energies.push_back(
-        values[values.size() - 2]);
-    m_current_wang_landau_system.max_boundaries_energies.push_back(
-        values[values.size() - 1]);
-  }
-
-  CV_minimum = *(std::min_element(
-      m_current_wang_landau_system.min_boundaries_energies.begin(),
-      m_current_wang_landau_system.min_boundaries_energies.end()));
-  CV_maximum = *(std::max_element(
-      m_current_wang_landau_system.max_boundaries_energies.begin(),
-      m_current_wang_landau_system.max_boundaries_energies.end()));
+    assert(values.size() >= 2);
+    auto it = values.end();
+    wl_system.max_boundaries_energies.push_back(*(--it));
+    wl_system.min_boundaries_energies.push_back(*(--it));
+  }
+
+  CV_minimum = *boost::range::min_element(wl_system.min_boundaries_energies);
+  CV_maximum = *boost::range::max_element(wl_system.max_boundaries_energies);
 }
 
 /**
  * Performs a randomly selected reaction in the reaction ensemble
  */
 int ReactionAlgorithm::do_reaction(int reaction_steps) {
   for (int i = 0; i < reaction_steps; i++) {
-    int reaction_id = i_random(reactions.size());
+    int reaction_id = i_random(static_cast<int>(reactions.size()));
```

Comment:
```c++
int ReactionAlgorithm::do_reaction(int reaction_steps) {
  printf("======\n");
  for (int i = 0; i < reaction_steps; i++) {
    int reaction_id = i_random(static_cast<int>(reactions.size()));
    printf("%d %zu\n", reaction_id, reactions.size());
    generic_oneway_reaction(reaction_id);
  }
  return 0;
}
```
output:
```
======
0 2
0 2
0 2
0 2
1 2
1 2
1 2
0 2
1 2
0 2
======
0 2
1 2
0 2
0 2
1 2
1 2
0 2
0 2
0 2
1 2
```

Reply:
What do you want to show here?

Reply:
ok I misunderstood your question. Do you want to pass a RNG lambda function as argument to `do_reaction()`?

Reply:
ah sorry, maybe it was not clear... yes, i wanted the RNG to be an input from outside of the function such that we could inject the randomness

Reply:
Maybe this can be done in the context of #4154? Right now injecting the randomness wouldn't help with unit testing, because RE classes still rely on the energy functions, that cannot be executed in a unit test.

Reply:
alright